### ISSUE TITLE
Add causal masking in attention_forward (j ≤ i)

### ISSUE THREAD

**Issue by sreehariam** [sreehariam/fa-research-cpp#5]
### Summary
Extend `fa::attention_forward` to support **causal attention** when `opts.causal == true`: each query at position i must only attend to keys j ≤ i.

### API
No signature change:
```
Tensor attention_forward(const Tensor& Q,
                         const Tensor& K,
                         const Tensor& V,
                         const Tensor* mask,          // optional padding mask (already supported)
                         const AttentionOpts& opts);  // NEW: honor opts.causal
```

### Behavior
- If `opts.causal == true`, set logits[j] = -inf for all j > i (before softmax).
- Padding mask (if provided) still applies in addition to causal (logical AND).
- Stable softmax (max-trick) as before.
- If all keys masked (by causal and/or padding), return a zero row.

### Acceptance
- Correct outputs on small deterministic cases (uniform over {0..i} when Q=K=0).
- Works for arbitrary B/H/N/D.
- Finite outputs; throws only for shape mismatches (not for causal).
- Does not regress existing P2P tests or PR1/PR2 tests.

### Out of scope
CUDA, alternate mask layouts.
